### DISCUSSION TITLE
How to deal with canceled checkout during first trial?

### DISCUSSION THREAD

**Discussion by guillaumebriday** [pay-rails/pay#827] [Q&A]
Hi,

I use [Stripe checkout](https://github.com/pay-rails/pay/blob/master/docs/stripe/8_stripe_checkout.md#stripe-checkout) and generic trial.

###  Code

On registration, I create a subscription with `fake_processor` to deal with trial without a payment method.

```ruby
# app/controllers/users/registrations_controller.rb
  def create
    super do
      if resource.persisted?
        trial_ends_at = TRIAL_PERIOD_DAYS.days.from_now

        resource.set_payment_processor(:fake_processor, allow_fake: true)
        resource.payment_processor.subscribe(trial_ends_at: trial_ends_at, ends_at: trial_ends_at)

        # ...
    end
  end
```

I got a controller to manage my subscriptions
```ruby
# app/controllers/app/subscriptions_controller.rb


  def checkout
    current_user.set_payment_processor(:stripe) # This is the important line

    checkout_session = current_user.payment_processor.checkout(
      mode: 'subscription',
      payment_method_types: ['card'],
      locale: I18n.locale,
      allow_promotion_codes: true,
      success_url: app_subscriptions_url,
      cancel_url: app_subscriptions_url,
      line_items: [{
        price: ENV.fetch('MONTHLY_PRICE_ID'),
        quantity: current_user.quantity_to_charge
      }],
      subscription_data: {
        trial_period_days: (current_user.trial_period_days if current_user.trial_period_days.positive?)
      }
    )

    redirect_to checkout_session.url, allow_other_host: true, status: :see_other
  end
```

## Issue

But if the customer does not complete the checkout and cancel it, the payment processor will still be `stripe` instead of `fake_processor` and the current trial period will be gone.

I don't really know how to deal with that without adding custom webhooks, so if you have a proper way to deal with it, that would be great!

Thanks in advance

**Comment by excid3** [comment]
Use `add_payment_processor :stripe` and then set `stripe` as the default after they've successfully checked out using the success_url redirect from Checkout.

**Reply by georgekettle** [reply]
Very helpful, thanks @excid3 

For anyone looking for implementation, here's what I did:

```
# GET /checkout/success
def success
    current_user.pay_customers.where(processor: "stripe").each do |pay_customer|
        pay_customer.sync_subscriptions(status: "all")
    end
    authorize current_user, :success?
    current_user.set_payment_processor(:stripe)
end
```

**Reply by georgekettle** [reply]
You'd probably only have 1 pay customer for each user, so you could simplify to this:
```
...
stripe_pay_customer = current_user.pay_customers.find_by(processor: "stripe")
stripe_pay_customer.sync_subscriptions(status: "all")
...
```

**Comment by guillaumebriday** [comment]
Hum, if I use `add_payment_processor` I got an error saying that `checkout` method does not exist.

<img width="1282" alt="Screenshot 2023-06-20 at 22 14 29" src="https://github.com/pay-rails/pay/assets/8252238/bf355581-fe48-4003-83a7-84803252f352">

It makes sense that it does not exist but I don't find anything in the docs related to that.

Am I missing something obvious? 🤔

**Comment by excid3** [comment]
Correct. `current_user.payment_processor` only returns the default. `add_payment_processor` adds and does not set a default.

```ruby
payment_processor = current_user.add_payment_processor(:stripe)
checkout_session = payment_processor.checkout(...)
```
